compiler/src/iree/compiler/Codegen/LLVMGPU/GPUTransformDialectStrategies.cpp

```diff
@@ -0,0 +1,423 @@
+///
+/// The tiling and distributing to blocks is done within a transform SequenceOp.
+/// It runs without interleaved canonicalize, CSE or enabling transforms which
+/// allows the transform dialect to build payload IR and not risk seeing it
+/// being DCE'd away immediately.
+static Value buildReductionStrategyBlockDistributionPart(
+    ImplicitLocOpBuilder &b, Value variantH, Value originalFillH,
+    Value originalGenericH, ArrayRef<int64_t> tileSizes0Generic) {
+  // Step 1. Split the reduction to get meatier parallelism.
+  // TODO: use a scf.foreach_thread for this.
+  auto splitReductionTransformOp =
+      b.create<SplitReductionOp>(originalGenericH,
+                                 /*splitFactor=*/2,
+                                 /*insertSplitDimension=*/1);
+  Value splitFillH = splitReductionTransformOp.getFillOp();
+  Value splitLinalgH = splitReductionTransformOp.getSplitLinalgOp();
+  Value combinerH = splitReductionTransformOp.getCombiningLinalgOp();
+  // TODO: IREE needs own workgroup mapping attribute.
+  // TODO: num of GPU block mapping attr is statically known here which is
+  // brittle. In the future, the builder of scf.foreach_thread can trim the
+  // number of mapping dims to the number of sizes.
+  auto x = mlir::gpu::GPUBlockMappingAttr::get(b.getContext(),
+                                               ::mlir::gpu::Blocks::DimX);
+  // Step 2. First level of tiling + fusion parallelizes to blocks using
+  // `tileSizes`.
+  tfdWithTileSizes<TileToForeachThreadAndWorkgroupCountRegionOp>(
+      b,
+      /*rootH=*/combinerH,
+      /*opsHToFuse=*/{originalFillH, splitFillH, splitLinalgH},
+      tileSizes0Generic, b.getArrayAttr({x}));
+
+  return variantH;
+}
+
+static Value buildReductionStrategyThreadDistributionPart(
+    ImplicitLocOpBuilder &b, Value variantH, ArrayRef<int64_t> tileSizes1Fill,
+    ArrayRef<int64_t> tileSizes1Generic) {
+  // TODO: Relying on ordering is brittle, harden this.
+  auto [fill2dH, parParRedH, fill1dH, parRedH] = matchAndUnpack<4>(
+      b, variantH,
+      ArrayRef<StringRef>{linalg::GenericOp::getOperationName(),
+                          linalg::FillOp::getOperationName()});
+  auto z = mlir::gpu::GPUThreadMappingAttr::get(b.getContext(),
+                                                ::mlir::gpu::Threads::DimZ);
+  auto y = mlir::gpu::GPUThreadMappingAttr::get(b.getContext(),
+                                                ::mlir::gpu::Threads::DimY);
+  (void)z;
+  (void)y;
+
+  // clang-format off
+  tfdWithTileSizes<TileToForeachThreadOp>(b,
+                   /*rootH=*/parRedH,
+                   /*opsHToFuse=*/{fill1dH},
+                  // TODO: activate this, we have the information but it does
+                  // not generate the IR we want atm.
+                  //  /*tileSizes=*/tileSizes1Fill,
+                  //  /*threadDimMapping=*/b.getArrayAttr({y}));
+                   /*tileSizes=*/ArrayRef<int64_t>{1, 0, 0},
```

Comment:
what's up with this?

Reply:
this is copy pasted, I have just moved code I didn't improve the matcher yet